Move torch/autograd/grad_mode.pyi stubs inline

rgommers opened the pull request:
- Add `torch._C` bindings from `torch/csrc/autograd/init.cpp`
- Renamed `torch._C.set_grad_enabled` to `torch._C._set_grad_enabled`
  so it doesn't conflict with torch.set_grad_enabled anymore

This is a continuation of gh-38201. All I did was resolve merge conflicts and finish the annotation of `_DecoratorContextManager.__call__` that @ezyang started in the first commit.

~Reverts commit b5cd3a80bbc, which was only motivated by not having `typing_extensions` available.~ (JIT can't be made to understand `Literal[False]`, so keep as is).

ezyang committed:
Move torch/autograd/grad_mode.pyi stubs inline

- Add torch._C bindings from torch/csrc/autograd/init.cpp
- Renamed torch._C.set_grad_enabled to torch._C._set_grad_enabled
  so it doesn't conflict with torch.set_grad_enabled anymore

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

(cherry picked from commit 551a222d663bc540bb037e690e7dad86f8f85af3)

rgommers committed:
Annotate the decorator usage of no_grad and enable_grad

SplitInfinity commented:
I am waiting for a final review of my JIT-side changes on #41371 before I can land it.

SplitInfinity commented:
#41371 has landed. Feel free to rebase and merge.

rgommers committed:
Merge branch 'master' into move-autograd-stub

rgommers committed:
Fix mismatching `__exit__` signatures in grad_mode.py

rgommers committed:
Return None from `__exit__()` method of `no_grad` et al after all

We can now rely on Literal from typing_extensions, but there's no
good way of making the JIT understand a `Literal[False]` return
value here (even when hiding the return type inside `if TYPE_CHECKING`).

So don't revert commit b5cd3a8 after all.
